reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -19,610 +19,1029 @@
+			return ((Fuseable.ConditionalSubscriber<? super T>) actual).tryOnNext(t);
+		}
+	}
+
+	static final class SwitchOnFirstControlSubscriber<T>
+			extends Operators.DeferredSubscription
+			implements InnerOperator<T, T>, ControlSubscriber<T> {
+
+		final AbstractSwitchOnFirstMain<?, T> parent;
+		final CoreSubscriber<? super T>       delegate;
+		final boolean                         cancelSourceOnComplete;
+
+		boolean done;
+
+		SwitchOnFirstControlSubscriber(AbstractSwitchOnFirstMain<?, T> parent,
+				CoreSubscriber<? super T> delegate,
+				boolean cancelSourceOnComplete) {
+			this.parent = parent;
+			this.delegate = delegate;
+			this.cancelSourceOnComplete = cancelSourceOnComplete;
+		}
+
+		@Override
+		public void sendSubscription() {
+			delegate.onSubscribe(this);
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (set(s)) {
+				long previousState = markOutboundSubscribed(this.parent);
+
+				if (hasOutboundCancelled(previousState)) {
+					s.cancel();
+				}
+			}
+		}
+
+		@Override
+		public CoreSubscriber<? super T> actual() {
+			return this.delegate;
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (this.done) {
+				Operators.onNextDropped(t, currentContext());
+				return;
+			}
+
+			this.delegate.onNext(t);
+		}
+
+		@Override
+		public void onError(Throwable throwable) {
+			if (this.done) {
+				Operators.onErrorDropped(throwable, currentContext());
+				return;
+			}
+
+			this.done = true;
+
+			final AbstractSwitchOnFirstMain<?, T> parent = this.parent;
+			long previousState = markOutboundTerminated(parent);
+
+			if (hasOutboundCancelled(previousState) || hasOutboundTerminated(previousState)) {
+				return;
+			}
+
+			if (!hasInboundCancelled(previousState) && !hasInboundTerminated(previousState)) {
+				parent.cancelAndError();
+			}
+
+			this.delegate.onError(throwable);
+		}
+
+		@Override
+		public void onComplete() {
+			if (this.done) {
+				return;
+			}
+
+			this.done = true;
+
+			final AbstractSwitchOnFirstMain<?, T> parent = this.parent;
+			long previousState = markOutboundTerminated(parent);
+
+			if (cancelSourceOnComplete && !hasInboundCancelled(previousState) && !hasInboundTerminated(previousState)) {
+				parent.cancelAndError();
+			}
+
+			this.delegate.onComplete();
+		}
+
+		@Override
+		public void cancel() {
+			REQUESTED.lazySet(this, STATE_CANCELLED);
+
+			final long previousState = markOutboundCancelled(this.parent);
+			if (hasOutboundCancelled(previousState) || hasOutboundTerminated(previousState)) {
+				return;
+			}
+
+			final boolean shouldCancelInbound =
+					!hasInboundTerminated(previousState) && !hasInboundCancelled(
+							previousState);
+
+			if (!hasOutboundSubscribed(previousState)) {
+				if (shouldCancelInbound) {
+					this.parent.cancel();
+				}
+				return;
+			}
+
+			this.s.cancel();
+
+			if (shouldCancelInbound) {
+				this.parent.cancelAndError();
+			}
+		}
+
+		@Override
+		public Object scanUnsafe(Attr key) {
```

Comment:
we could technically answer the `Attr.CANCELLED` attribute here, wdyt?
(both here and in the conditional variant)

Reply:
fixed